<a href="https://colab.research.google.com/github/payel462/ML-PROJECTS/blob/main/NLP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#import io
from google.colab import drive
drive.mount("/content/drive/")

Mounted at /content/drive/


## Split the content whenever you get "."(full stop)

In [ ]:
import pandas as pd
#df=pd.read_csv('/content/drive/MyDrive/Math Data (7500) final.xlsx - Sheet1.csv')
#df=pd.read_excel('/content/drive/MyDrive/Math Data (7500) final.xlsx')
df=pd.read_excel('/content/drive/MyDrive/Merged_BothDS_Dataset.xlsx')
#df=open(r'/content/drive/MyDrive/Math Data (7500) final.xlsx - Sheet1.csv','r')
#contents=df.read()
#li=contents.split("|?")
#print(contents)
df.head()

Problem   Equation Sollution  \
0  সানির কিছু মার্বেল রয়েছে। তিনি ৪৮৯ মার্বেল কি...   ২৭৭৮-৪৮৯      1929   
1  কমলির ১২১ টি খেলনা ছিল। এখন তার ২২ টি খেলনা আছ...     ১২১-২২        99   
2  কমলিনির কিছু মার্বেল ছিল। সে জুয়ানকে ১৮৩ গুলি...    ৫৯৩-১৮৩       410   
3  কনিকার ২৩১৫ টি লাল কলম এবং ১০২৮ টি নীল কলম রয়...  ২৩১৫+১০২৮      3343   
4  ইশিতার ৩৪৪ টি বই রয়েছে। ২৩৬ টি গল্পের বই এবং ...    ৩৪৪-২৩৬       108   

  Class  
0     -  
1     -  
2     -  
3     +  
4     -

In [ ]:
#li = contents.split()
#print(li)

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## Ultimately You have to store every words or tokens into a list without all the puctuations and special charactors.

In [ ]:
'''import string

def clean_text(doc):
  tokens= doc.split()
  table= str.maketrans('', '', string.punctuation)
  tokens =[w.translate(table) for w in tokens]
  tokens= [word for word in tokens if word.isalpha()]
  tokens= [word.lower() for word in tokens]
  return tokens'''

"import string\n\ndef clean_text(doc):\n  tokens= doc.split()\n  table= str.maketrans('', '', string.punctuation)\n  tokens =[w.translate(table) for w in tokens]\n  tokens= [word for word in tokens if word.isalpha()]\n  tokens= [word.lower() for word in tokens]\n  return tokens"

In [ ]:
#tokens = clean_text(contents)
#print(tokens)

In [ ]:
!pip3 install bnltk

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Preparing metadata (setup.py) ... done
  Created wheel for sklearn: filename=sklearn-0.0.post1-py3-none-any.whl size=2344 sha256=62a709e2524bff942215d759631774164cfe27ac1e3c3865469107a7e49babb0
  Stored in directory: /root/.cache/pip/wheels/14/25/f7/1cc0956978ae479e75140219088deb7a36f60459df242b1a72
Successfully built sklearn


In [ ]:
# Get tokens and remove punctuation

#import string
#from bnltk.tokenize import Tokenizers

#t = Tokenizers()
#tokens = t.bn_word_tokenizer(contents)


In [ ]:
#print(tokens)

##28.06.2022


In [ ]:
df['Problem'][1]

'কমলির ১২১ টি খেলনা ছিল। এখন তার ২২ টি খেলনা আছে, তাহলে সে কতগুলো খেলনা হারিয়ে ফেলেছে?'

In [ ]:
# Get tokens and remove punctuation

import string
from bnltk.tokenize import Tokenizers

t = Tokenizers()
problem_lines=[]

for line in df['Problem']:
  tokens = t.bn_word_tokenizer(line)
  problem_lines.append(tokens)

problem_lines[0:5]

[['সানির',
  'কিছু',
  'মার্বেল',
  'রয়েছে',
  'তিনি',
  '৪৮৯',
  'মার্বেল',
  'কিনেছিলেন',
  'এখন',
  'তার',
  'সব',
  'মিলিয়ে',
  '২৭৭৮',
  'মার্বেল',
  'রয়েছে',
  'তিনি',
  'কয়টি',
  'দিয়ে',
  'শুরু',
  'করেছিলেন'],
 ['কমলির',
  '১২১',
  'টি',
  'খেলনা',
  'ছিল',
  'এখন',
  'তার',
  '২২',
  'টি',
  'খেলনা',
  'আছে',
  'তাহলে',
  'সে',
  'কতগুলো',
  'খেলনা',
  'হারিয়ে',
  'ফেলেছে'],
 ['কমলিনির',
  'কিছু',
  'মার্বেল',
  'ছিল',
  'সে',
  'জুয়ানকে',
  '১৮৩',
  'গুলি',
  'মার্বেল',
  'দিয়েছে',
  'এখন',
  'তার',
  'হাতে',
  'রয়েছে',
  '৫৯৩',
  'টি',
  'মার্বেল',
  'প্রথমে',
  'তার',
  'কতগুলি',
  'মার্বেল',
  'ছিল'],
 ['কনিকার',
  '২৩১৫',
  'টি',
  'লাল',
  'কলম',
  'এবং',
  '১০২৮',
  'টি',
  'নীল',
  'কলম',
  'রয়েছে',
  'তার',
  'মোট',
  'কতটি',
  'কলম',
  'রয়েছে'],
 ['ইশিতার',
  '৩৪৪',
  'টি',
  'বই',
  'রয়েছে',
  '২৩৬',
  'টি',
  'গল্পের',
  'বই',
  'এবং',
  'বাকিগুলি',
  'খেলাধুলার',
  'বিষয়ে',
  'তবে',
  'খেলা',
  'সম্পর্কিত',
  'কয়টি',
  'বই',
  'রয়েছে',
  'ইশার']]

##how to convert sequence of token to sequence of integer using tokenizer

In [ ]:
import pandas as pd
#df=pd.read_csv('/content/drive/MyDrive/Math Data (7500) final.xlsx - Sheet1.csv')
df=pd.read_excel('/content/drive/MyDrive/Math Data (7500) final.xlsx')

df.head(5)

Problem   Equation Sollution  \
0  সানির কিছু মার্বেল রয়েছে। তিনি ৪৮৯ মার্বেল কি...   ২৭৭৮-৪৮৯    1929.0   
1  কমলির ১২১ টি খেলনা ছিল। এখন তার ২২ টি খেলনা আছ...     ১২১-২২      99.0   
2  কমলিনির কিছু মার্বেল ছিল। সে জুয়ানকে ১৮৩ গুলি...    ৫৯৩-১৮৩     410.0   
3  কনিকার ২৩১৫ টি লাল কলম এবং ১০২৮ টি নীল কলম রয়...  ২৩১৫+১০২৮    3343.0   
4  ইশিতার ৩৪৪ টি বই রয়েছে। ২৩৬ টি গল্পের বই এবং ...    ৩৪৪-২৩৬     108.0   

  Class  
0     -  
1     -  
2     -  
3     +  
4     -

In [ ]:
!pip3 install bnltk

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
# Get tokens and remove punctuation

import string
from bnltk.tokenize import Tokenizers

t = Tokenizers()
problem_lines=[]

for line in df['Problem']:
  tokens = t.bn_word_tokenizer(line)
  problem_lines.append(tokens)

problem_lines[0:2]

[['সানির',
  'কিছু',
  'মার্বেল',
  'রয়েছে',
  'তিনি',
  '৪৮৯',
  'মার্বেল',
  'কিনেছিলেন',
  'এখন',
  'তার',
  'সব',
  'মিলিয়ে',
  '২৭৭৮',
  'মার্বেল',
  'রয়েছে',
  'তিনি',
  'কয়টি',
  'দিয়ে',
  'শুরু',
  'করেছিলেন'],
 ['কমলির',
  '১২১',
  'টি',
  'খেলনা',
  'ছিল',
  'এখন',
  'তার',
  '২২',
  'টি',
  'খেলনা',
  'আছে',
  'তাহলে',
  'সে',
  'কতগুলো',
  'খেলনা',
  'হারিয়ে',
  'ফেলেছে']]

In [ ]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.preprocessing.text import Tokenizer

In [ ]:
tokenizer = Tokenizer()
tokenizer.fit_on_texts(problem_lines)

In [ ]:
sequence=tokenizer.texts_to_sequences(problem_lines)

In [ ]:
sequence[0]

[1341,
 71,
 66,
 3,
 13,
 1717,
 66,
 132,
 14,
 9,
 54,
 55,
 1859,
 66,
 3,
 13,
 17,
 75,
 102,
 97]

In [ ]:
word_index = tokenizer.word_index
print(word_index)

{'টি': 1, 'কত': 2, 'রয়েছে': 3, 'টাকা': 4, 'আছে': 5, 'মোট': 6, 'কাছে': 7, 'এবং': 8, 'তার': 9, 'ছিল': 10, 'একটি': 11, 'করে': 12, 'তিনি': 13, 'এখন': 14, 'কতগুলি': 15, 'তবে': 16, 'কয়টি': 17, '১০': 18, 'জন': 19, 'প্রতিটি': 20, 'আরও': 21, 'মাইল': 22, '৫': 23, '৪': 24, 'গাছ': 25, 'থেকে': 26, '২০': 27, '২': 28, 'বই': 29, 'যদি': 30, '৩': 31, 'সে': 32, 'থাকবে': 33, '৬': 34, '৫০': 35, 'আর': 36, '৮': 37, 'দাম': 38, 'হবে': 39, 'করেছে': 40, 'আপেল': 41, 'জন্য': 42, 'বেলুন': 43, '৭': 44, 'প্রতি': 45, 'বাক্সে': 46, '৩০': 47, 'ঘন্টা': 48, 'তোমার': 49, 'তারা': 50, 'পেন্সিল': 51, 'কলম': 52, 'করতে': 53, 'সব': 54, 'মিলিয়ে': 55, '৯': 56, 'ঝিনুক': 57, '০': 58, '৪০': 59, 'হলে': 60, '১': 61, 'কমলালেবু': 62, 'তাকে': 63, 'কতগুলো': 64, 'ও': 65, 'মার্বেল': 66, 'বিস্কুট': 67, 'ডিম': 68, 'বাক্স': 69, 'বেশি': 70, 'কিছু': 71, '১৫': 72, 'করেছেন': 73, 'ভাগ': 74, 'দিয়ে': 75, 'বেগে': 76, 'লেবু': 77, 'আজ': 78, 'টুকরো': 79, 'কলা': 80, 'তাদের': 81, 'থাকে': 82, 'মাছ': 83, 'চেয়ে': 84, 'তাস': 85, 'লজেন্স': 86, '২৫': 87, 'বছ

In [ ]:
word_index

{'টি': 1,
 'কত': 2,
 'রয়েছে': 3,
 'টাকা': 4,
 'আছে': 5,
 'মোট': 6,
 'কাছে': 7,
 'এবং': 8,
 'তার': 9,
 'ছিল': 10,
 'একটি': 11,
 'করে': 12,
 'তিনি': 13,
 'এখন': 14,
 'কতগুলি': 15,
 'তবে': 16,
 'কয়টি': 17,
 '১০': 18,
 'জন': 19,
 'প্রতিটি': 20,
 'আরও': 21,
 'মাইল': 22,
 '৫': 23,
 '৪': 24,
 'গাছ': 25,
 'থেকে': 26,
 '২০': 27,
 '২': 28,
 'বই': 29,
 'যদি': 30,
 '৩': 31,
 'সে': 32,
 'থাকবে': 33,
 '৬': 34,
 '৫০': 35,
 'আর': 36,
 '৮': 37,
 'দাম': 38,
 'হবে': 39,
 'করেছে': 40,
 'আপেল': 41,
 'জন্য': 42,
 'বেলুন': 43,
 '৭': 44,
 'প্রতি': 45,
 'বাক্সে': 46,
 '৩০': 47,
 'ঘন্টা': 48,
 'তোমার': 49,
 'তারা': 50,
 'পেন্সিল': 51,
 'কলম': 52,
 'করতে': 53,
 'সব': 54,
 'মিলিয়ে': 55,
 '৯': 56,
 'ঝিনুক': 57,
 '০': 58,
 '৪০': 59,
 'হলে': 60,
 '১': 61,
 'কমলালেবু': 62,
 'তাকে': 63,
 'কতগুলো': 64,
 'ও': 65,
 'মার্বেল': 66,
 'বিস্কুট': 67,
 'ডিম': 68,
 'বাক্স': 69,
 'বেশি': 70,
 'কিছু': 71,
 '১৫': 72,
 'করেছেন': 73,
 'ভাগ': 74,
 'দিয়ে': 75,
 'বেগে': 76,
 'লেবু': 77,
 'আজ': 78,
 'টুকরো': 79,
 'কলা': 80,
 'তাদের'

##Make sequence of integers for each problems

##For an example, [1341, 71, 66, 3, 13, 1717, 66, 132, 14, 9, 54, 55, 1859, 66, 3, 13, 17, 75, 102, 97] will be ur output for a problem.

padding of test data

In [ ]:
from tensorflow.keras.preprocessing.sequence import pad_sequences
max_length = 55
lines_pad = pad_sequences(sequence, maxlen = max_length, padding ='post')
lines_pad[0]

array([1341,   71,   66,    3,   13, 1717,   66,  132,   14,    9,   54,
         55, 1859,   66,    3,   13,   17,   75,  102,   97,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0],
      dtype=int32)

In [ ]:
#shuffle the dataset
#mapp target value with one hot encoding

In [ ]:
#categorical to numerical
df["Class"] = df["Class"].map({' -':0, ' +':1,'*':2,'/':3, '+':1})

In [ ]:
df.head()

Problem   Equation Sollution  \
0  সানির কিছু মার্বেল রয়েছে। তিনি ৪৮৯ মার্বেল কি...   ২৭৭৮-৪৮৯    1929.0   
1  কমলির ১২১ টি খেলনা ছিল। এখন তার ২২ টি খেলনা আছ...     ১২১-২২      99.0   
2  কমলিনির কিছু মার্বেল ছিল। সে জুয়ানকে ১৮৩ গুলি...    ৫৯৩-১৮৩     410.0   
3  কনিকার ২৩১৫ টি লাল কলম এবং ১০২৮ টি নীল কলম রয়...  ২৩১৫+১০২৮    3343.0   
4  ইশিতার ৩৪৪ টি বই রয়েছে। ২৩৬ টি গল্পের বই এবং ...    ৩৪৪-২৩৬     108.0   

   Class  
0      0  
1      0  
2      0  
3      1  
4      0

In [ ]:
df = df.sample(frac = 1)  
df

Problem    Equation Sollution  \
1299  চারটি ঝুড়িতে আম আছে। প্রথম ঝুড়িতে ১৯ টি, দ্বিত...  ১৯+১৯+৮+১৪      60.0   
5884  অনুপমের ৯ টি পেন্সিল রয়েছে। অনুপমকে আরও ১০ টি...        ১০+৯      19.0   
3788  পার্কে বর্তমানে ৪১ টি ছোট গাছ এবং ৪৪ টি লম্বা ...     ৪১ + ৫৭      98.0   
7136  অনুষ্ঠানে ২৫০ জন অতিথি আসছেন। প্রতিটি টেবিল ২৫...      ২৫০/২৫      10.0   
6583  একটি কেক তৈরি করতে ৪ টি আপেল লাগে। ২০০  টি কেক...       ২০০/৪      50.0   
...                                                 ...         ...       ...   
1094  একটি কেকের রেসিপিতে ক্রিমের জন্য ০.৬ কাপ চিনি ...     ০.৬+০.২       0.8   
6373  তিস্তা যদি ২ ঘন্টায় ৬ মাইল বেগে দৌড়ে যায়।তবে...         ২*৬      12.0   
527   একটি বাসে ৪২ জন যাত্রী ছিল । পথে ২৩ জন নেমে গে...       ৪২-২৩      19.0   
4551  সুমন এর কাছে ১ টি ল্যাপটপ  আছে তার দাদার কাছে ...       ১+১+১       3.0   
532   একটি বই রাখার আলমারিতে ৭ টি তাক আছে । একটি তাক...        ১২*৭      84.0   

      Class  
1299      1  
5884      1  
3788      1  
7136      3  
6583      3  
...     ...  
1094      1  
6373      2  
527       0  
4551      1  
532       2  

[7666 rows x 4 columns]

In [ ]:

#from numpy import argmax
#char_to_int = dict((c, i) for i, c in enumerate(word_index)
#integer_encoded = [char_to_int[char] for char in word_index]
#print(integer_encoded)
# one hot encode
#onehot_encoded = list()
#for value in integer_encoded:
#	letter = [0 for _ in range(len(sequence[0]))]
#	letter[value] = 1
#	onehot_encoded.append(letter)
#print(onehot_encoded)
# invert encoding
#inverted = int_to_char[argmax(onehot_encoded[0])]
#print(inverted)

In [ ]:
from numpy import array
from numpy import argmax
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder
# define example
#data =word_index
#values = array(data)
#label_encoder = LabelEncoder()
#integer_encoded = label_encoder.fit_transform(class)
#print(integer_encoded)
# binary encode
#onehot_encoder = OneHotEncoder(sparse=False)
#integer_encoded = integer_encoded.reshape(len(integer_encoded), 1)
#onehot_encoded = onehot_encoder.fit_transform(integer_encoded)
#print(onehot_encoded)
# invert first example
#inverted = label_encoder.inverse_transform([argmax(onehot_encoded[0, :])])
#print(inverted)

In [ ]:
# Shuffle our dataset
import numpy as np
category = df["Class"].values

indices = np.arange(lines_pad.shape[0])
np.random.shuffle(indices)
lines_pad = lines_pad[indices]
category = category[indices]

In [ ]:
category

array([1, 0, 0, ..., 1, 1, 0])

In [ ]:
#from sklearn.model_selection import train_test_split
#X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=2)

In [ ]:
#Convert the classes into one hot vector
n_values = np.max(category) + 1
y = np.eye(n_values)[category]

In [ ]:
y[0]

array([0., 1., 0., 0.])

In [ ]:
lines_pad[0]

array([2486, 1603, 1056,  183,  109,  659, 1896,   27,    4, 1775,  676,
       1057, 1897, 3073,    4, 1775,  676, 1898,  756, 1056,    6,    2,
          4,  103,   40,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
          0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0],
      dtype=int32)

In [ ]:
# Split the dataset
#manually splitting
split_data = 0.2
num_split_samples = int(split_data*lines_pad.shape[0])

x = lines_pad[:-num_split_samples]
y = y[ :-num_split_samples]
x_test = lines_pad[-num_split_samples:]
y_test = y[-num_split_samples:]

In [ ]:
num_split_samples #test data

1533

In [ ]:
y

array([[0., 1., 0., 0.],
       [1., 0., 0., 0.],
       [1., 0., 0., 0.],
       ...,
       [0., 0., 0., 1.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.]])

lstm model building

In [ ]:
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import LSTM
from keras.layers import Dropout

In [ ]:
Lstm_model= Sequential()

Lstm_model.add(LSTM(units = 50, return_sequences = True, input_shape = (x.shape[1], 1)))
Lstm_model.add(Dropout(0.2))

Lstm_model.add(LSTM(units = 50, return_sequences = True))
Lstm_model.add(Dropout(0.2))

Lstm_model.add(LSTM(units = 50, return_sequences = True))
Lstm_model.add(Dropout(0.2))

Lstm_model.add(LSTM(units = 50))
Lstm_model.add(Dropout(0.2))

Lstm_model.add(Dense(units = 4))

Lstm_model.compile(optimizer = 'adam', loss = 'mean_squared_error')

In [ ]:
#Build Model
from keras.models import Sequential
from keras.layers import Dense, Embedding, LSTM
#from keras.layers.embeddings import Embedding
from tensorflow.keras.layers import Embedding
embedding_dim = 100

print('Build model.....')

model= Sequential()
model.add(Embedding(len(word_index)+1, embedding_dim, input_length=max_length))
model.add(LSTM(units=32, dropout=0.5, recurrent_dropout=0.2))
model.add(Dense(4, activation='softmax'))

model.compile(loss='categorical_crossentropy', optimizer= 'adam', metrics=['accuracy'])
model.summary()

Build model.....
Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 55, 100)           451700    
                                                                 
 lstm_4 (LSTM)               (None, 32)                17024     
                                                                 
 dense_1 (Dense)             (None, 4)                 132       
                                                                 
Total params: 468,856
Trainable params: 468,856
Non-trainable params: 0
_________________________________________________________________


In [ ]:
print('Train....')

history=model.fit(x, y, batch_size=32, epochs=100, validation_data=(x_test, y_test))

Train....
Epoch 1/100
192/192 [==============================] - 18s 73ms/step - loss: 1.3212 - accuracy: 0.3608 - val_loss: 1.3089 - val_accuracy: 0.3705
Epoch 2/100
192/192 [==============================] - 14s 71ms/step - loss: 1.3170 - accuracy: 0.3631 - val_loss: 1.3092 - val_accuracy: 0.3705
Epoch 3/100
192/192 [==============================] - 13s 69ms/step - loss: 1.3168 - accuracy: 0.3651 - val_loss: 1.3112 - val_accuracy: 0.3705
Epoch 4/100
192/192 [==============================] - 14s 73ms/step - loss: 1.3160 - accuracy: 0.3651 - val_loss: 1.3099 - val_accuracy: 0.3705
Epoch 5/100
192/192 [==============================] - 13s 70ms/step - loss: 1.3162 - accuracy: 0.3651 - val_loss: 1.3099 - val_accuracy: 0.3705
Epoch 6/100
192/192 [==============================] - 14s 72ms/step - loss: 1.3161 - accuracy: 0.3651 - val_loss: 1.3093 - val_accuracy: 0.3705
Epoch 7/100
192/192 [==============================] - 13s 70ms/step - loss: 1.3159 - accuracy: 0.3651 - val_loss: 1.309

KeyboardInterrupt: ignored

In [ ]:
x.shape

In [ ]:
history=Lstm_model.fit(x,y, epochs = 100, batch_size = 32)

In [ ]:
import tensorflow as tf
from tensorflow.keras.datasets import imdb
from tensorflow.keras.layers import Embedding, Dense, LSTM
from tensorflow.keras.losses import BinaryCrossentropy
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.preprocessing.sequence import pad_sequences

# Model configuration
additional_metrics = ['accuracy']
batch_size = 128
embedding_output_dims = 15
loss_function = BinaryCrossentropy()
max_sequence_length = 300
num_distinct_words = 5000
number_of_epochs = 5
optimizer = Adam()
validation_split = 0.20
verbosity_mode = 1

# Disable eager execution
tf.compat.v1.disable_eager_execution()

# Load dataset
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=num_distinct_words)
print(x_train.shape)
print(x_test.shape)

# Pad all sequences
padded_inputs = pad_sequences(x_train, maxlen=max_sequence_length, value = 0.0) # 0.0 because it corresponds with <PAD>
padded_inputs_test = pad_sequences(x_test, maxlen=max_sequence_length, value = 0.0) # 0.0 because it corresponds with <PAD>

# Define the Keras model
model = Sequential()
model.add(Embedding(num_distinct_words, embedding_output_dims, input_length=max_sequence_length))
model.add(LSTM(10))
model.add(Dense(1, activation='sigmoid'))

# Compile the model
model.compile(optimizer=optimizer, loss=loss_function, metrics=additional_metrics)

# Give a summary
model.summary()

# Train the model
history = model.fit(padded_inputs, y_train, batch_size=batch_size, epochs=number_of_epochs, verbose=verbosity_mode, validation_split=validation_split)

# Test the model after training
test_results = model.evaluate(padded_inputs_test, y_test, verbose=False)
print(f'Test results - Loss: {test_results[0]} - Accuracy: {100*test_results[1]}%')

In [ ]:
add=0
sub=0
mul=0
div=0
for i in range(0,8664):
  if df["Class"][i]=='+':
    add=add+1
  elif df["Class"][i]=='-':
    sub=sub+1
  elif df["Class"][i]=='/':
    div=div+1
  elif df["Class"][i]=='*':
    mul=mul+1
  elif df["Class"][i]==' +':
    add=add+1
  elif df["Class"][i]==' -':
    sub=sub+1
  elif df["Class"][i]==' /':
    div=div+1
  elif df["Class"][i]==' *':
    mul=mul+1



In [ ]:
print(add)
print(sub)
print(mul)
print(div)


3201
2508
1927
1027


In [ ]:
df["Equation"][i]

'২৫০+৫০+৩০০'

In [ ]:
len(df["Equation"][i])



10

#ADD n0+n1+n2

In [ ]:
add=0
for i in range(0,8664):
  count=0
  print(df["Equation"][i])
  for j in (df["Equation"][i]):
    if j=='+':
      count=count+1
  if count==2:
    add=add+1


২৭৭৮-৪৮৯
১২১-২২
৫৯৩-১৮৩
২৩১৫+১০২৮
৩৪৪-২৩৬
২২৩+৯৫
২৭*১৮
৫-৪.২৮
৩*১.৫৮
৪৮/৪
২৫০*৫
৫২৫৯+২৯৮৭
২৭*৪
৩৪০৮-১৪৫
৪৫*৯
১২৯*৩
৩৪*৯
১৯*৫
১২৭-৮৮
৩৯*১০
৯*২
৬৫৮৮৯+৮৮৭৬
৩৭৫/৩
১৭*৪
১৪৪*৪
১৪৪*৩
৩০*৪
২৮০-২১
১৫৯*৮
৩৬০+৮৫
৬১০-৩২৭
৭৫০-৫৩২
২৩*৪৫
১৫৬*১২
১০*৭
১৮*১৯
১২*৩৮
৫৫*৩৬
৪*৫১০
১২*৩৮
১২০/১০
৮*২
৪২/৩
৭৫/৩
৪*৮
৬০*২৪
৫৫*৮
৩৮*৩
৩৪*৩
১০*১৭
১০*৩৩
২৩*২
১১*৮
২০*১৪
৪*৮
১৫*৩
৩৯*২৬৯
২৫৬০/১০
২৬৫০/১০
৪৫৫/৫
২০৮-১৭
১১৫*১২
১৫২৪০/৮
২১৯/৩
২২৪/৮
৮৬০-৩৬৭
১২০৯+২১৮৬
৪০৩+২৮৭
৪০৩+২৮৭
১৪*৬
১৫৬+১৮৭
৫৬৩-১৪৭
২১৯+২২৯
৮৭৩-৫৬৩
৯৯৮-৬৭৯
৯০০-১৫৬
৩৭২-২১২.৩
২৪৩৬/৩
০.৩৮-০.০৫
৭+২
০.৬৮-০.২৫
৭+৫
২৫৮-৭৫
৫০০-৭৯
০.৯-০.৪
২৬.৩*৫
৩*৩
১৫*১০
১৫+২০+৩
১৮৬+৭৫
৪*১৫
৬*৪
১২৭-৯৯
৬৯-৩০
৯৭-৫০
৮৮+১১০
১০৮+৫৩+২২
৫৮+৭০+২৯
৭৯+১০১
১০৯+৫০+৫০
৮+৮
১২/৪
১৭+৫০+২০
৮+৮
১২÷৪
১৭+৫০+২০
৩৯+৪৮
১২৭×২
৪০÷৫
৮ টা ৩০ মিনিট -৮ টা
১৩৫×৫
১২০÷৬
৬৬÷৬
৭০০÷৭
৬+৮
১৯+১৯+৮+১৪
২৯-২০
৪৯×৩
২৯÷৩
৩০*৩
২৫৭×১২
৬০.৫৬+৮.৪০+৯০.৫০+২০.৫০
৭০-৬০
৮৫৮÷১৭
১০৮-(২০.২৫+১৫.৭৫)
৯৮-৮৯
৭+১৩
৮৯÷৪
২২+১৮+২৪
(১০×৮)+৮
(৬৮×৪) - ২০৫
৭৮ - ২৯
৫৭৭ - (২৪৫+২০৫)
১০৮+৫৫
১০৮×৯
৯২-৮০
৩৯-১০
(৩০০ ÷ ২) × ৮
৪০/২০
৫ × (৩০ × ৩)
৬৮০ × ১২
৪

TypeError: ignored

In [ ]:
503
111

